src/libraries/System.Private.CoreLib/src/System/Guid.cs

```diff
@@ -261,6 +272,54 @@ public Guid(string g)
             this = result.ToGuid();
         }
 
+        /// <summary>Gets the value of the variant field for the <see cref="Guid" />.</summary>
+        /// <remarks>
+        ///     <para>This corresponds to the most significant 4 bits of the 8th byte: 00000000-0000-0000-F000-000000000000. The "don't-care" bits are not masked out.</para>
+        ///     <para>See RFC 9562 for more information on how to interpret this value.</para>
+        /// </remarks>
+        public int Variant => _d >> 4;
+
+        /// <summary>Gets the value of the version field for the <see cref="Guid" />.</summary>
+        /// <remarks>
+        ///     <para>This corresponds to the most significant 4 bits of the 6th byte: 00000000-0000-F000-0000-000000000000.</para>
+        ///     <para>See RFC 9562 for more information on how to interpret this value.</para>
+        /// </remarks>
+        public int Version => _c >>> 12;
+
+        /// <summary>Creates a new <see cref="Guid" /> according to RFC 9562, following the Version 7 format.</summary>
+        /// <returns>A new <see cref="Guid" /> according to RFC 9562, following the Version 7 format.</returns>
+        /// <remarks>
+        ///     <para>This uses <see cref="DateTimeOffset.UtcNow" /> to determine the Unix Epoch timestamp source.</para>
+        ///     <para>This seeds the rand_a and rand_b sub-fields with random data.</para>
+        /// </remarks>
+        public static Guid CreateVersion7() => CreateVersion7(DateTimeOffset.UtcNow);
+
+        /// <summary>Creates a new <see cref="Guid" /> according to RFC 9562, following the Version 7 format.</summary>
+        /// <param name="timestamp">The date time offset used to determine the Unix Epoch timestamp</param>
+        /// <returns>A new <see cref="Guid" /> according to RFC 9562, following the Version 7 format.</returns>
+        /// <remarks>
+        ///     <para>This seeds the rand_a and rand_b sub-fields with random data.</para>
+        /// </remarks>
+        public static Guid CreateVersion7(DateTimeOffset timestamp)
+        {
+            // This isn't the most optimal way, but we don't have an easy way to get
+            // secure random bytes in corelib without doing this, and its not terrible
+            Guid result = NewGuid();
+
+            // 2^48 is roughly 8925.5 years, which from the Unix Epoch means we won't
+            // overflow until around July of 10,895. So there isn't much need to handle it
+            // particularly given that DateTimeOffset.MaxValue is December 31, 9999
+            long unix_ts_ms = timestamp.ToUnixTimeMilliseconds();
+
+            Unsafe.AsRef(in result._a) = (int)(unix_ts_ms >> 16);
```

Comment:
We don't need to worry about sign extension here because unix_ts_ms will never be negative?

Reply:
Hmmm, I think it actually can be negative and we may need to detect this case. In particular, `DateTimeOffset` supports timestamps prior to the Unix Epoch and so can represent `Jan 1st, 0001`, meaning `ToUnixTimeMilliseconds` would return `0xFFFFC77CEDD32800` (`-62135596800000`).

The RFC specifies
> UUIDv7 features a time-ordered value field derived from the widely implemented and well-known Unix Epoch timestamp source, the number of milliseconds since midnight 1 Jan 1970 UTC, leap seconds excluded.
and
> 48-bit big-endian unsigned number of the Unix Epoch timestamp in milliseconds

So negative timestamps would be an error case.